### PR TITLE
Add classes for AuthnContext structure

### PR DESCRIPTION
**Pull request by tvdijen**
Adds classes + unit tests for:
AuthnContext
AuthnContextDecl
AuthenticatingAuthority

Implements these classes into the codebase

### CODE REVIEW

**Comment by tvdijen** [comment]
Just a quick sanity check will suffice

File: src/SAML2/XML/saml/AuthnContextDecl.php [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+/**
+ * Class representing SAML2 AuthnContextDecl
+ *
+ * @author Tim van Dijen, <tvdijen@gmail.com>
+ * @package SimpleSAMLphp
+ */
+class AuthnContextDecl extends \SAML2\XML\AbstractConvertable
+{
+    /** @var \DOMElement|null */
```
[the remark this answers is not in the document]
Comment (jaimeperez) [reply src/SAML2/XML/saml/AuthnContextDecl.php]: Actually, I was wondering... would it make sense to use `\SAML2\XML\Chunk` in here, instead of `DOMElement` directly?

File: src/SAML2/XML/saml/AuthnContext.php [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+     * @param \SAML2\XML\saml\AuthenticatingAuthority[]|null $authenticatingAuthorities
+     */
+    public function __construct(
+        ?AuthnContextClassRef $authnContextClassRef,
+        ?AuthnContextDecl $authnContextDecl,
+        ?AuthnContextDeclRef $authnContextDeclRef,
+        ?array $authenticatingAuthorities
+    ) {
+        if (!is_null($authnContextClassRef)) {
+            Assert::oneOf(
```
Comment (jaimeperez) [inline src/SAML2/XML/saml/AuthnContext.php]: Does `oneOf` allow the case of both being `null`? Because in this case, the minimum occurs is set to zero, so both could be skipped in fact.

File: src/SAML2/XML/saml/AuthnContext.php [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+        } else {
+            Assert::false(
+                is_null($authnContextDecl) && is_null($authnContextDeclRef),
+                'You need either an AuthnContextDecl or an AuthnContextDeclRef'
+            );
+        }
+
+        $this->setAuthnContextClassRef($authnContextClassRef);
+
+        if (!is_null($authnContextDecl)) {
```
Comment (jaimeperez) [inline src/SAML2/XML/saml/AuthnContext.php]: Given that the setters have nullable parameters, we can skip these `is_null()` checks and just call the setters.

File: src/SAML2/XML/saml/AuthnContext.php [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+    }
+
+
+    /**
+     * Set the value of the authnContextClassRef-property
+     *
+     * @param \SAML2\XML\saml\AuthnContextClassRef|null $authnContextClassRef
+     * @return void
+     */
+    public function setAuthnContextClassRef(?AuthnContextClassRef $authnContextClassRef): void
```
Comment (jaimeperez) [inline src/SAML2/XML/saml/AuthnContext.php]: I would also mark this as protected at least. The simpler the interface, the better, so if we set everything in the constructor, there's no point in having public setters, IMHO.

File: src/SAML2/XML/saml/AuthnContext.php [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+
+    /**
+     * Set the value of the authenticatingAuthorities-property
+     *
+     * @param \SAML2\XML\saml\AuthenticatingAuthority[]|null $authenticatingAuthorities
+     * @return void
+     */
+    public function setAuthenticatingAuthorities(?array $authenticatingAuthorities): void
+    {
+        $this->authenticatingAuthorities = $authenticatingAuthorities;
```
Comment (jaimeperez) [inline src/SAML2/XML/saml/AuthnContext.php]: Here I think we should check the elements and verify that all of them are `SAML2\XML\saml\AuthenticatingAuthority` objects.

File: src/SAML2/XML/saml/AuthnContext.php [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+                [],
+                [$authnContextDecl, $authnContextDeclRef],
+                'Can only have one of AuthnContextDecl/AuthnContextDeclRef'
+            );
+        } else {
+            Assert::false(
+                empty($authnContextDecl) && empty($authnContextDeclRef),
+                'You need either an AuthnContextDecl or an AuthnContextDeclRef'
+            );
+        }
```
Comment (jaimeperez) [inline src/SAML2/XML/saml/AuthnContext.php]: We also have this exact check in the constructor. Can we skip it here, considering we're just passing the processed values to it later in the method?

File: src/SAML2/XML/saml/AuthnContext.php [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+
+        /** @var \DOMElement[] $authenticatingAuthorities */
+        $authenticatingAuthorities = Utils::xpQuery($xml, './saml_assertion:AuthenticatingAuthority');
+
+        return new self(
+            empty($authnContextClassRef) ? null : AuthnContextClassRef::fromXML($authnContextClassRef[0]),
+            empty($authnContextDecl) ? null : AuthnContextDecl::fromXML($authnContextDecl[0]),
+            empty($authnContextDeclRef) ? null : AuthnContextDeclRef::fromXML($authnContextDeclRef[0]),
+            array_map([AuthenticatingAuthority::class, 'fromXML'], $authenticatingAuthorities) ?: null
```
Comment (jaimeperez) [inline src/SAML2/XML/saml/AuthnContext.php]: 👏

File: src/SAML2/XML/saml/AuthnContext.php [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+                null,
+                [$this->authnContextDecl, $this->authnContextDeclRef],
+                'Can only have one of AuthnContextDecl/AuthnContextDeclRef'
+            );
+        } else {
+            Assert::false(
+                empty($this->authnContextDecl) && empty($this->authnContextDeclRef),
+                'You need either an AuthnContextDecl or an AuthnContextDeclRef'
+            );
+        }
```
Comment (jaimeperez) [inline src/SAML2/XML/saml/AuthnContext.php]: If the setters are protected (or private), then every possible path to create an object will lead to this check in the constructor. Then the check becomes redundant here.

File: tests/SAML2/XML/saml/AuthenticatingAuthorityTest.php [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+/**
+ * Class \SAML2\XML\saml\AuthenticatingAuthorityTest
+ */
+class AuthenticatingAuthorityTest extends \PHPUnit\Framework\TestCase
+{
+    /**
+     * @return void
+     */
+    public function testMarshalling(): void
```
Comment (jaimeperez) [inline tests/SAML2/XML/saml/AuthenticatingAuthorityTest.php]: With the new `AbstractConvertable` class, we can now simplify tests like this by casting the object to a string and comparing with a string representing the XML document.

File: src/SAML2/XML/saml/AuthenticatingAuthority.php [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    /**
+     * Set the value of the authority-property
+     *
+     * @param string $name
+     * @return void
+     */
+    public function setAuthority(string $authority): void
+    {
+        $authority = trim($authority);
+        Assert::stringNotEmpty($authority);
```
[the remark this answers is not in the document]
Comment (tvdijen) [reply src/SAML2/XML/saml/AuthenticatingAuthority.php]: That's a good catch.. I'll remove the assertion and double-check the already merged PRs

File: src/SAML2/XML/saml/AuthenticatingAuthority.php [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+
+    /**
+     * Set the value of the authority-property
+     *
+     * @param string $name
+     * @return void
+     */
+    public function setAuthority(string $authority): void
```
[the remark this answers is not in the document]
Comment (tvdijen) [reply src/SAML2/XML/saml/AuthenticatingAuthority.php]: That's exactly what I was suggesting in #189.. We should do that everywhere and make the objects immutable, but that's a design choice I can't make on my own..

File: src/SAML2/XML/saml/AuthnContext.php (hunk 4 above)
Comment (tvdijen) [reply src/SAML2/XML/saml/AuthnContext.php]: OK, if that's what we agree upon, then I'm going to change that for the already merged PR's as well

File: src/SAML2/XML/saml/AuthenticatingAuthority.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (jaimeperez) [reply src/SAML2/XML/saml/AuthenticatingAuthority.php]: I think the approach to make objects immutable would help us reduce complexity and have easier to use interfaces, so let's go ahead 👍

File: src/SAML2/XML/saml/AuthnContextDecl.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (tvdijen) [reply src/SAML2/XML/saml/AuthnContextDecl.php]: I think so, but where would we do the conversion? In the constructor? In the calling code?

[the remark this answers is not in the document]
Comment (jaimeperez) [reply src/SAML2/XML/saml/AuthnContextDecl.php]: Intuitively, I'd say in the calling code. Then, if your code supports a specific declaration, you can always model that extending `Chunk`, and still pass it around here...

File: src/SAML2/XML/saml/AuthnContext.php (hunk 2 above)
Comment (tvdijen) [reply src/SAML2/XML/saml/AuthnContext.php]: Yes, that is allowed. I've double checked by adding an `Assert::oneOf(null, [null, null], 'This is wrong');` and it didn't trigger the error message..

Comment (jaimeperez) [reply src/SAML2/XML/saml/AuthnContext.php]: Perfect!

File: src/SAML2/XML/saml/AuthnContext.php (hunk 7 above)
Comment (tvdijen) [reply src/SAML2/XML/saml/AuthnContext.php]: Yeah, that was quite interesting :P

File: src/SAML2/XML/saml/AuthnContextDecl.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (tvdijen) [reply src/SAML2/XML/saml/AuthnContextDecl.php]: I've pushed the requested changes.. Can you give it another look?

File: src/SAML2/XML/saml/AuthnContextDecl.php [hunk 12]
```diff
@@ -68,7 +67,7 @@ public function setDecl(DOMElement $decl): void
      */
     public static function fromXML(DOMElement $xml): object
     {
-        return new self($xml);
+        return new self(new Chunk($xml));
```
Comment (jaimeperez) [inline src/SAML2/XML/saml/AuthnContextDecl.php]: I'm not sure about this one. What you would pass to the `fromXML()` methods are the elements themselves, right? In this case, the `<saml:AuthnContextDecl>` element. The chunk in this case would then be the contents inside `$xml`.

File: src/SAML2/XML/saml/AuthenticatingAuthority.php [hunk 13]
```diff
@@ -69,6 +66,9 @@ public function setAuthority(string $authority): void
      */
     public static function fromXML(DOMElement $xml): object
     {
+        Assert::same($xml->tagName, 'saml:AuthenticatingAuthority');
+        Assert::same($xml->namespaceURI, Constants::NS_SAML);
```
Comment (jaimeperez) [inline src/SAML2/XML/saml/AuthenticatingAuthority.php]: Great! These checks should also be applied to the other elements, I think.

File: src/SAML2/XML/saml/AuthenticatingAuthority.php [hunk 14]
```diff
@@ -92,7 +90,7 @@ public function toXML(DOMElement $parent = null): DOMElement
             $parent->appendChild($e);
         }
 
-        $e->textContent = $this->authority;
+        $e->textContent = $this->getAuthority();
```
Comment (jaimeperez) [inline src/SAML2/XML/saml/AuthenticatingAuthority.php]: Why this change?

File: src/SAML2/XML/saml/AuthenticatingAuthority.php (hunk 13 above)
Comment (tvdijen) [reply src/SAML2/XML/saml/AuthenticatingAuthority.php]: I will do that

File: src/SAML2/XML/saml/AuthnContextDecl.php (hunk 12 above)
Comment (tvdijen) [reply src/SAML2/XML/saml/AuthnContextDecl.php]: Hmm, then how is it possible for the tests to pass?

File: src/SAML2/XML/saml/AuthenticatingAuthority.php (hunk 14 above)
Comment (tvdijen) [reply src/SAML2/XML/saml/AuthenticatingAuthority.php]: My bad... I dunno what I was thinking, but it's undone

File: src/SAML2/XML/saml/AuthenticatingAuthority.php (hunk 13 above)
Comment (tvdijen) [reply src/SAML2/XML/saml/AuthenticatingAuthority.php]: I've added a commit for AuthnContext.. Can't do AuthnContextDecl because that one can contains virtually anything

File: src/SAML2/XML/saml/AuthnContextDecl.php (hunk 12 above)
Comment (tvdijen) [reply src/SAML2/XML/saml/AuthnContextDecl.php]: Hmm, I see I do pass the elements themselves..

Comment (tvdijen) [reply src/SAML2/XML/saml/AuthnContextDecl.php]: I guess you've discovered a bug.. I've commit a fix

**Comment by tvdijen** [comment]
If you could give it another go please? All comments have been addressed and I added a commit that fixes the already merged PR's

File: src/SAML2/XML/saml/AuthnContextClassRef.php [hunk 15]
```diff
@@ -69,6 +65,9 @@ public function setClassRef(string $classRef): void
      */
     public static function fromXML(DOMElement $xml): object
     {
+        Assert::same($xml->tagName, 'saml:AuthnContextClassRef');
```
Comment (jaimeperez) [inline src/SAML2/XML/saml/AuthnContextClassRef.php]: This should check `$xml->localName` instead. The namespace prefix can be anything, the standard does not mandate it to be `saml`.

Comment (tvdijen) [reply src/SAML2/XML/saml/AuthnContextClassRef.php]: That goes for all of them I guess?

Comment (jaimeperez) [reply src/SAML2/XML/saml/AuthnContextClassRef.php]: Yep, we should always use `localName` when comparing.

File: src/SAML2/XML/saml/AuthnContextDecl.php [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+     * @param \DOMElement $xml The XML element we should load
+     * @return \SAML2\XML\saml\AuthnContextDecl
+     */
+    public static function fromXML(DOMElement $xml): object
+    {
+        Assert::same($xml->tagName, 'saml:AuthnContextDecl');
+        Assert::same($xml->namespaceURI, Constants::NS_SAML);
+
+        /** @psalm-var \DOMElement $xml->childNodes[1] */
+        return new self(new Chunk($xml->childNodes[1]));
```
Comment (jaimeperez) [inline src/SAML2/XML/saml/AuthnContextDecl.php]: I was already thinking about a problem with this before seeing your implementation, and now I see it confirmed 😞. The contents of this element can be anything, meaning these two would be valid:

```xml
<saml:AuthnContextDecl>
  <foo>bar</foo>
</saml:AuthnContextDecl>
```

```xml
<saml:AuthnContextDecl>
  <foo>1</foo>
  <bar>2</bar>
</saml:AuthnContextDecl>
```

Both are valid, but the current implementation would only work with the former. The latter would yield an object that contains only the `<foo>` element. Even worse, this would also be valid:

```xml
<saml:AuthnContextDecl>
  foo
  <bar>2</bar>
</saml:AuthnContextDecl>
```

I don't see an easy way out, but I draw some conclusions:

1. We can't assume the number of nodes children of the element to be only 1.
2. We can't use `SAML2\XML\Chunk` with the contents, because that's intended for XML elements, but the contents can be of any type, and in an arbitrary number.

I think probably the best solution would be to recursively import all the child nodes into a new `DOMDocument` and append them there. The `DOMDocument` would then be the property we save, and the class needs to implement `\Serializable` the same way `SAML2\XML\Chunk` does it.

Comment (tvdijen) [reply src/SAML2/XML/saml/AuthnContextDecl.php]: Now I see why nobody ever implemented AuthnContextDecl and StatusDetail :rofl:

File: src/SAML2/XML/samlp/Status.php [hunk 17]
```diff
@@ -66,7 +67,7 @@ public function getStatusCode(): StatusCode
      *
      * @throws \InvalidArgumentException if assertions are false
      */
-    public function setStatusCode(StatusCode $statusCode): void
+    private function setStatusCode(StatusCode $statusCode): void
```
Comment (jaimeperez) [inline src/SAML2/XML/samlp/Status.php]: I'm not completely convinced we should declare all setters as private. In some cases, I think it would be desirable to extend an existing class to implement improved functionality, and then access to those methods would be needed. Do we already have any classes under `SAML2\XML` that extend others?

File: src/SAML2/XML/samlp/StatusDetail.php [hunk 18]
```diff
@@ -64,7 +65,11 @@ public function setDetail(?\DOMElement $detail): void
      */
     public static function fromXML(DOMElement $xml): object
     {
-        return new self($xml);
+        Assert::same($xml->tagName, 'samlp:StatusDetail');
+        Assert::same($xml->namespaceURI, Constants::NS_SAMLP);
+
+        /** @psalm-var \DOMElement $xml->childNodes[1] */
+        return new self(new Chunk($xml->childNodes[1]));
```
Comment (jaimeperez) [inline src/SAML2/XML/samlp/StatusDetail.php]: Having a look at the core document of the standard, here we also have a problem with the contents. The good news, in this case, is that the contents are supposed to be XML elements indeed, so we could use an array of `SAML2\XML\Chunk` objects.

File: tests/SAML2/XML/saml/AuthenticatingAuthorityTest.php [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+    {
+        $authority = new AuthenticatingAuthority('https://sp.example.com/SAML2');
+
+        $this->assertEquals(
+            strval($authority),
+            '<saml:AuthenticatingAuthority xmlns:saml="' . Constants::NS_SAML
+                . '">https://sp.example.com/SAML2</saml:AuthenticatingAuthority>'
+        );
```
Comment (jaimeperez) [inline tests/SAML2/XML/saml/AuthenticatingAuthorityTest.php]: A minor detail here: the parameters to `assertEquals()` are the expected value and the actual value, in that order, so we should probably invert them here to avoid confusions in case of an error.

File: src/SAML2/XML/saml/AuthnContextDecl.php (hunk 16 above)
Comment (jaimeperez) [reply src/SAML2/XML/saml/AuthnContextDecl.php]: 😆

Comment (tvdijen) [reply src/SAML2/XML/saml/AuthnContextDecl.php]: https://stackoverflow.com/questions/5389076/difference-similarities-between-xsdany-and-xsdanytype
I think the 3rd case is out of the question, so that would leave us with an array of Chunks here as well

File: src/SAML2/XML/samlp/Status.php (hunk 17 above)
Comment (tvdijen) [reply src/SAML2/XML/samlp/Status.php]: We extend from SignedElementHelper, RoleDescriptor, SSODescriptorType, EndpointType, NameIDType and Attribute..

I can only imagine doing so if you want to support some almost-but-not-really-SAML spec.. Strictly speaking the classes should be final, but I guess that's pretty harsh..

Comment (jaimeperez) [reply src/SAML2/XML/samlp/Status.php]: I actually like the approach to make classes final better than making setters private. In the end, the standard has its own extension points, where you should add your functionality, whatever that shall be. We should follow the same principle, and the rest of the classes representing elements defined by SAML2 should probably be final as you are saying, unless we have a reason not to make them so (e.g. we're extending them somewhere, but in that case we're talking about types, rather than actual elements, so for those it wouldn't make sense to mark them as final.

I'm wondering if we should use traits instead of an abstract class that we are extending...

File: src/SAML2/XML/saml/AuthnContextDecl.php (hunk 16 above)
Comment (tvdijen) [reply src/SAML2/XML/saml/AuthnContextDecl.php]: Should be done.. Please take a look

File: src/SAML2/XML/samlp/Status.php (hunk 17 above)
Comment (tvdijen) [reply src/SAML2/XML/samlp/Status.php]: If we make them public again, we have to add a lot of protecting assertions to toXML again to avoid illegal situations.. I thought we had agreed on the immutable objects..
I'd say final+private and/or non-final+protected..

File: src/SAML2/XML/samlp/StatusDetail.php (hunk 18 above)
Comment (tvdijen) [reply src/SAML2/XML/samlp/StatusDetail.php]: I read this a but too late, so I took the same approach as AuthnContextDecl.. A later commit reverts this and uses Chunks..

File: src/SAML2/XML/AbstractConvertable.php [hunk 20]
```diff
@@ -27,6 +29,36 @@ public function __toString(): string
...
 
 
+    /**
+     * Serialize this XML chunk
+     *
+     * @return string The serialized chunk.
+     */
+    public function serialize(): string
+    {
+        return serialize($this->toXML()->ownerDocument->saveXML());
```
Comment (jaimeperez) [inline src/SAML2/XML/AbstractConvertable.php]: There's no need to call `serialize()` here. The `serialize()` method of the `Serializable` interface needs to return a string that you can later recover, but that doesn't mean it needs to follow any particular format. Returning the XML document as a string should suffice, as we can recover it later from that representation.

File: src/SAML2/XML/AbstractConvertable.php [hunk 21]
```diff
@@ -27,6 +29,36 @@ public function __toString(): string
...
+     * Un-serialize this XML chunk.
+     *
+     * @param string $serialized The serialized chunk.
+     * @return void
+     *
+     * Type hint not possible due to upstream method signature
+     */
+    public function unserialize($serialized): void
+    {
+        $doc = DOMDocumentFactory::fromString(unserialize($serialized));
```
Comment (jaimeperez) [inline src/SAML2/XML/AbstractConvertable.php]: ... and the same here. No need to call `unserialize()`.

File: src/SAML2/XML/saml/AuthenticatingAuthority.php [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+    public function toXML(DOMElement $parent = null): DOMElement
+    {
+        if ($parent === null) {
+            $doc = DOMDocumentFactory::create();
+            $e = $doc->createElementNS(Constants::NS_SAML, 'saml:AuthenticatingAuthority');
+            $doc->appendChild($e);
+        } else {
+            $e = $parent->ownerDocument->createElementNS(Constants::NS_SAML, 'saml:AuthenticatingAuthority');
+            $parent->appendChild($e);
+        }
```
Comment (jaimeperez) [inline src/SAML2/XML/saml/AuthenticatingAuthority.php]: This block is a pattern that repeats a lot. It might be an idea to move it to a generic method inside `AbstractConvertable` receiving the namespace and the name of the element we want to create, plus the parent. Then all these blocks of identical code can be substituted with a single call to that method.

File: src/SAML2/XML/saml/AuthnContextDecl.php [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+
+    /**
+     * Un-serialize this XML chunk.
+     *
+     * @param string $serialized The serialized chunk.
+     * @return void
+     *
+     * Type hint not possible due to upstream method signature
+     */
+    public function unserialize($serialized): void
```
Comment (jaimeperez) [inline src/SAML2/XML/saml/AuthnContextDecl.php]: This shouldn't be necessary, I believe. The parent calling `fromXML()` should be enough to recreate the object, right?

File: src/SAML2/XML/AbstractConvertable.php (hunk 20 above)
Comment (tvdijen) [reply src/SAML2/XML/AbstractConvertable.php]: Done

File: src/SAML2/XML/AbstractConvertable.php (hunk 21 above)
Comment (tvdijen) [reply src/SAML2/XML/AbstractConvertable.php]: And done

File: src/SAML2/XML/samlp/Status.php (hunk 17 above)
Comment (jaimeperez) [reply src/SAML2/XML/samlp/Status.php]: No, no, I didn't mean making them public, but protected. But of course that would only make sense for non-final classes. Also, provided that the setters are protected, I'm fine with the properties being private, since any class potentially extending can still get and set those values.

File: src/SAML2/XML/samlp/StatusCode.php [hunk 24]
```diff
@@ -97,6 +97,9 @@ public function setSubCodes(?array $subCodes): void
      */
     public static function fromXML(DOMElement $xml): object
     {
+        Assert::same($xml->localName, 'StatusCode');
+        Assert::same($xml->namespaceURI, Constants::NS_SAMLP);
+
```
Comment (jaimeperez) [inline src/SAML2/XML/samlp/StatusCode.php]: I can't add a comment on the proper line, so I'll add it here. Line 147:

```php
        if (!empty($this->subCodes)) {
            foreach ($this->subCodes as $subCode) {
                $subCode->toXML($e);
            }
        }
```

The outer if can be removed I think, since if `$this->subCodes` is empty, the loop will not iterate.

File: src/SAML2/XML/samlp/StatusDetail.php [hunk 25]
```diff
@@ -18,41 +20,73 @@
...
+        $this->details = $details;
+    }
+
+
+    /**
+     * Serialize this XML chunk
+     *
+     * @return string The serialized chunk.
+     */
+    public function serialize(): string
```
Comment (jaimeperez) [inline src/SAML2/XML/samlp/StatusDetail.php]: `serialize()` and `unserialize()` shouldn't be necessary here either. `fromXML()` should handle reconstruction of the object.

File: src/SAML2/XML/samlp/StatusDetail.php [hunk 26]
```diff
@@ -64,7 +98,11 @@ public function setDetail(?\DOMElement $detail): void
      */
     public static function fromXML(DOMElement $xml): object
     {
-        return new self($xml);
+        Assert::same($xml->localName, 'StatusDetail');
+        Assert::same($xml->namespaceURI, Constants::NS_SAMLP);
+
+        /** @psalm-var \DOMElement $xml->childNodes[1] */
+        return new self([new Chunk($xml->childNodes[1])]);
```
Comment (jaimeperez) [inline src/SAML2/XML/samlp/StatusDetail.php]: This is a bug. We should iterate over the `DOMNodeList` in `$xml->childNodes` and create a new `SAML2\XML\Chunk` for each `DOMElement`, then pass an array of all those chunks to the constructor.

File: src/SAML2/XML/samlp/StatusCode.php (hunk 24 above)
Comment (tvdijen) [reply src/SAML2/XML/samlp/StatusCode.php]: It's optional, so it could be null..

File: src/SAML2/XML/samlp/StatusDetail.php [hunk 27]
```diff
@@ -85,8 +123,10 @@ public function toXML(DOMElement $parent = null): DOMElement
             $parent->appendChild($e);
         }
 
-        if (!empty($this->detail)) {
-            $e->appendChild($e->ownerDocument->importNode($this->detail, true));
+        if (!empty($this->details)) {
```
Comment (jaimeperez) [inline src/SAML2/XML/samlp/StatusDetail.php]: Oh, I can actually comment on this one 😅

Same case as before, the `if` is redundant in this case, as `foreach` tests the condition.

File: src/SAML2/XML/samlp/StatusDetail.php (hunk 25 above)
Comment (tvdijen) [reply src/SAML2/XML/samlp/StatusDetail.php]: Removed

File: tests/SAML2/XML/saml/AuthenticatingAuthorityTest.php [hunk 28]
```diff
@@ -0,0 +1,48 @@
...
+
+/**
+ * Class \SAML2\XML\saml\AuthenticatingAuthorityTest
+ */
+class AuthenticatingAuthorityTest extends \PHPUnit\Framework\TestCase
+{
+    /**
+     * @return void
+     */
+    public function testMarshalling(): void
```
Comment (jaimeperez) [inline tests/SAML2/XML/saml/AuthenticatingAuthorityTest.php]: These tests look wonderful now, so simple and easy to read! ❤️

File: src/SAML2/XML/samlp/StatusCode.php (hunk 24 above)
Comment (jaimeperez) [reply src/SAML2/XML/samlp/StatusCode.php]: My bad, didn't notice it could be null!

File: src/SAML2/XML/samlp/StatusDetail.php (hunk 26 above)
Comment (tvdijen) [reply src/SAML2/XML/samlp/StatusDetail.php]: Has been resolved
